File: backend-burgers/src/models/Afspraak.ts
import {objectType} from "nexus";
import {Context} from "../context";

const Afspraak = objectType({
	name: "Afspraak",
	definition: t => {
		t.int("id", {
			description: "Een unique identifier voor een afspraak in het systeem.",
		});
		t.string("omschrijving");
		t.string("bedrag");
		t.boolean("credit");
		t.field("betaalinstructie", {
			type: "Betaalinstructie",
			resolve: (root: any) => {
				const betaalinstructie = root.betaalinstructie;

				if (!betaalinstructie) {
					return null;
				}

				return {
					byDay: betaalinstructie.by_day,
					byMonth: betaalinstructie.by_month,
					byMonthDay: betaalinstructie.by_month_day,
					repeatFrequency: betaalinstructie.repeat_frequency,
					exceptDates: betaalinstructie.except_dates,
					startDate: betaalinstructie.start_date,
					endDate: betaalinstructie.end_date,
				};
			},
		});
		t.string("validFrom", {
			resolve: root => root["valid_from"],
		});
		t.string("validThrough", {
			resolve: root => root["valid_through"],
		});
		t.field("tegenrekening", {
			type: "Rekening",
			resolve: async (root, _, ctx: Context) => {
				const rekeningId = root["tegen_rekening_id"];

				if (!rekeningId) {
					return null;
				}

				return await ctx
					.dataSources.huishoudboekjeservice
					.getRekeningById(rekeningId);
			},
		});
		t.list.field("journaalposten", {
			type: "Journaalpost",
			resolve: async (root, _, ctx: Context) => {
				const {id} = root;

				if (!id) {
					return [];
				}

				return await ctx
					.dataSources.huishoudboekjeservice
					.getJournaalpostenByAfspraakId(id);
			},
		});
	},
});

export default Afspraak;

File: backend/hhb_backend/service/model/afspraak.py
from typing import List, Optional

from hhb_backend.graphql.scalars.day_of_week import DayOfWeekEnum
from hhb_backend.service.model.base_model import BaseModel


class Betaalinstructie(BaseModel):
    by_day: Optional[List[DayOfWeekEnum]]
    by_month: Optional[List[int]]
    by_month_day: Optional[List[int]]
    repeat_frequency: Optional[str]  # unused
    except_dates: Optional[List[str]]
    start_date: Optional[str]
    end_date: Optional[str]


class Afspraak(BaseModel):
    id: int
    omschrijving: str
    bedrag: int
    credit: bool
    valid_from: str
    valid_through: Optional[str]
    rubriek_id: int
    zoektermen: Optional[List[str]]
    burger_id: int
    afdeling_id: Optional[int]
    postadres_id: Optional[str]
    alarm_id: Optional[str]
    tegen_rekening_id: int
    betaalinstructie: Optional[Betaalinstructie]
    aantal_betalingen: Optional[int]  # unused
    # exposed relations
    journaalposten: List[int]
    overschrijvingen: List[int]
